plutus-core/plutus-core/src/Data/SatInt.hs

```diff
@@ -246,21 +265,52 @@ quotRemSI a@(I# _) b@(I# _) = (SI (a `quotInt` b), SI (a `remInt` b))
 divModSI ::  Int -> Int -> (SatInt, SatInt)
 divModSI x@(I# _) y@(I# _) = (SI (x `divInt` y), SI (x `modInt` y))
 
+
 plusSI :: SatInt -> SatInt -> SatInt
+plusSI (SI (I# x#)) (SI (I# y#)) =
+  case addIntC# x# y#  of
+    (# r#, f# #) ->
+        if isTrue# f# then  -- Overflow;  I think the value of f# is always 1 in this case
+                            -- and doesn't give us any other useful information
+        if      isTrue# ((x# ># 0#) `andI#` (y# ># 0#)) then maxBound
+        else if isTrue# ((x# <# 0#) `andI#` (y# <# 0#)) then minBound
```

Comment:
I don't understand this: the documentation of `addIntC` says that `r#` is the sum truncated to fit in an Int... isn't that exactly what we want? I can see from your experiments that that's not working, but it seems very odd.

Reply:
I now think we should be very careful about how we use `isTrue#`.  The documentation for `addInt#` says

> Add signed integers reporting overflow. First member of result is the sum truncated to an Int#; second member is zero if the true sum fits in an Int#, nonzero if overflow occurred (the sum is either too large or too small to fit in an Int#).

and for `isTrue#` it says

> Alias for tagToEnum#. Returns True if its parameter is 1# and False if it is 0#.

For the code above we've got `f#` and we want to branch if it's nonzero, but the documentation for `isTrue#` says nothing about what happens if its argument isn't `0#` or `1#`, and indeed the behaviour is weird:
```
> isTrue# 0#
False
> isTrue# 1#
True
> isTrue# 2#
False
> isTrue# 3#
True
> isTrue# 4#
True
> isTrue# 5#
False
> isTrue# 6#
True
> isTrue# 7#
True
> isTrue# 8#
True
> isTrue# 9#
True
> isTrue# 10#
True
> isTrue# 11#
False
> isTrue# 12#
False
> isTrue# 13#
False
> isTrue# 14#
False
> isTrue# 15#
False
> isTrue# 16#
False
```
Experiments with `addInt#` and `subInt#` suggests that the second component of the return value is always `0#` or `1#`, so this shouldn't matter; however all it says is that the second component is nonzero in case of overflow, and for all we know the exact value may depend on the CPU architecture or something.  Unfortunately `isTrue#` seems to be the only efficient way to check values;  there's not even an `isFalse#` (but see the note `Note [Optimizing isTrue#]` in [this file](https://hackage.haskell.org/package/ghc-prim-0.3.1.0/docs/src/GHC-Types.html)).

I think that to be on the safe side we really have to do something like 
```
if isTrue# (f# ==# 0#) then ...
```
to be safe (although I couldn't find an explicit statement of what `==#` is supposed to do.

Also we should probably say
```
if      isTrue# (x# ># 0#)  && isTrue# (y# ># 0#) then maxBound
```
instead of trying to be fancy with `and#`.  We'll only run this if there's an overflow, so it doesn't matter if it's not too efficient.

Reply:
> I don't understand this: the documentation of `addIntC` says that `r#` is the sum truncated to fit in an Int... isn't that exactly what we want? I can see from your experiments that that's not working, but it seems very odd.

I think maybe it's truncated to 64 bits, not a (signed) `Int`.  The maximum value of `Int#` is 2^63 - 1 = 9223372036854775807,
and if we put a=1111111111111111111 then 9*a is bigger than this (and so overflows) but still fits into 64 bits, as a negative number.  

Up above, we had a+8*a = -8446744073709551617 (but 9*a wasn't equal to this!), and as integers 2^64-8446744073709551617 = 9999999999999999999, so I think there's some two's complement stuff going on.  I always get very confused by this stuff though.

The documentation for `addIntC#` says

> First member of result is the sum truncated to an Int#; 

and I think that must mean that what it's doing is equivalent to adding the numbers as integers and then forgetting everything except the bottom 64 bits (although I'm not sure how that explanation would work if negative numbers were involved).

Next it says

> second member is zero if the true sum fits in an Int#, nonzero if overflow occurred (the sum is either too large or too small to fit in an Int#).

I think it should really say "fit in a _signed_ Int'` there.

Reply:
Yikes. Okay, then I agree we need to do something like you suggest.

Reply:
> Experiments with addInt# and subInt# suggests that the second component of the return value is always 0# or 1#, so this shouldn't matter; however all it says is that the second component is nonzero in case of overflow, and for all we know the exact value may depend on the CPU architecture or something.

GHC itself has [tests](https://gitlab.haskell.org/ghc/ghc/-/blob/master/testsuite/tests/primops/should_run/T9430.hs#L65) for `addIntC` that seem to assume that it's indeed 0 or 1, so I think we can assume that. I'll ask in #ghc anyway.